src/common/io-config/src/s3_provider.rs

```diff
@@ -0,0 +1,86 @@
+use std::hash::{Hash, Hasher};
+
+use pyo3::{
+    types::{PyBytes, PyModule},
+    PyAny, PyObject, PyResult, Python, ToPyObject,
+};
+use serde::{de::Visitor, Serialize};
+
+#[derive(Clone, Debug)]
+pub struct S3CredentialsProvider {
+    pub provider: PyObject,
+    pub hash: isize,
+}
+
+impl S3CredentialsProvider {
+    pub fn new(py: Python, provider: &PyAny) -> PyResult<Self> {
```

Comment:
you should just do `provider.to_object(py)` outside of this call and just pass in the pyobj

Reply:
I still need py to get the hash so it might be cleaner to just pass in these two values